repository/gridftp/globus5/authz/source/Config.c

```
@@ -0,0 +1,497 @@
+static bool config_loaded = false;
+
+// Initialize the read-write lock for the configuration object
+pthread_rwlock_t config_rwlock = PTHREAD_RWLOCK_INITIALIZER;
+
+/******************************************************************************
+ * File Scoped Functions
+ ******************************************************************************/
+
+/**
+ * @brief Sets a value to a destination buffer with length checks.
+ *
+ * This function copies a source string (`a_src`) to a destination buffer
+ * (`a_dest`) while ensuring the length of the source string does not exceed the
+ * maximum allowed length
+ * (`a_max_len`). If the source string is empty or too long, an error is logged,
+ * and the function returns `true`. Otherwise, the source string is copied to
+ * the destination.
+ *
+ * @param a_label The label associated with the configuration value, used in
+ * error logs.
+ * @param a_dest A pointer to the destination buffer where the source string
+ * will be copied.
+ * @param a_src A pointer to the source string to be copied to the destination
+ * buffer.
+ * @param a_max_len The maximum allowable length for the destination buffer.
+ *
+ * @return `true` if an error occurred (e.g., empty source or source too long),
+ * `false` otherwise.
+ *
+ * @note The function uses `strncpy` to copy the source string into the
+ * destination buffer, ensuring that no more than `a_max_len` characters are
+ * copied. If the source string is longer than the allowed maximum length, an
+ * error is logged, and the value is not copied.
+ */
+bool setConfigValInternal(const char *a_label, char *a_dest, const char *a_src,
+                          size_t a_max_len) {
+  size_t len = strlen(a_src);
+  if (len == 0) {
+    AUTHZ_LOG_ERROR("DataFed - '%s' value not set.\n", a_label);
+    return true;
+  }
+
+  if (len > a_max_len) {
+    AUTHZ_LOG_ERROR("DataFed - '%s' value too long (max %zu).\n", a_label,
+                    a_max_len);
+    return true;
+  }
+

```

General C question from this.

```c
 * This function copies a source string (`a_src`) to a destination buffer
 * (`a_dest`) while ensuring the length of the source string does not exceed the
 * maximum allowed length
```

Is it bad practice to ensure null termination when using  `strncpy`, even if we know we don't exactly need it? There's no scenario where `a_src` is longer than `a_max_len` cause
```c
if (len > a_max_len)
```

So no buffer overruns or memory leaks.

I was reading this https://cboard.cprogramming.com/c-programming/117061-null-termination-str-n-cpy-strlen.html#post872229

---

This is a really good catch, and demonstrates your attention to detail. You are correct, strlen only returns the number of chars excluding the null pointer. This should raise an error if len + 1 > a_max_len. strncpy will auto append \0 as long as there is room.

---

Yeah this is a great catch, I didn't think about the possibility of running out of space for the null terminator during copy.
I'm not sure if it's more clear to write `len + 1 > a_max_len` or `len >= a_max_len` or `size_t len = strlen(a_src) + 1;` but whichever way I feel like it should be noted somewhere nearby the reason for a change.